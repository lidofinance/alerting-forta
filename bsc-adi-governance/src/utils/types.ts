import { FindingSeverity, FindingType, ethers } from 'forta-agent'
import { Log } from '@ethersproject/abstract-provider'

export type EventOfNotice = {
  address: string
  event: string
  alertId: string
  name: string
  description: CallableFunction
  severity: FindingSeverity
  type: FindingType
}

export type TransactionDto = {
  logs: Log[]
  to: string | null
  block: {
    timestamp: number
    number: number
  }
}

export type LogDescription = ethers.utils.LogDescription
