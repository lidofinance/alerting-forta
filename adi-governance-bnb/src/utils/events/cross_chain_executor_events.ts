import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { EventOfNotice } from '../types'

export function getCrossChainExecutorEvents(CROSS_CHAIN_EXECUTOR: string): EventOfNotice[] {
  const uniqueKeys = [
    'e39b460a-69fa-40ff-b1c7-e90aaedd707d',
    'eb9108cc-5d06-450b-87ed-a3c962e0e77f',
    '980e06e4-e745-4d06-af7a-0a830e8d9520',
    '6413fc86-b592-4ea1-8e0d-530d222b1309',
    '5d9c4f9e-70a4-4784-8a41-aebac1165135',
    '4bb06b1c-6b47-4000-89b4-a2fc577ba812',
    '017edc0c-e3b4-43da-96ed-c8c247bab8c1',
    '227da709-90e6-4e0f-9eaf-3944baab5ea0',
  ]

  return [
    {
      address: CROSS_CHAIN_EXECUTOR,
      event: 'event GuardianUpdate(address oldGuardian, address newGuardian)',
      alertId: 'BNB-ADI-GUARDIAN-UPDATED',
      name: '🚨🚨🚨  BNB a.DI: Guardians updated',
      description: (args: Result) => `Guardian was updated from ` + `${args.oldGuardian} to ${args.newGuardian}`,
      severity: FindingSeverity.Critical,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
    {
      address: CROSS_CHAIN_EXECUTOR,
      event: 'event DelayUpdate(uint256 oldDelay, uint256 newDelay)',
      alertId: 'BNB-ADI-DELAY-UPDATED',
      name: '⚠️ BNB a.DI: Cross-chain executor delay updated',
      description: (args: Result) => `Delay was updated from ` + `${args.oldDelay} to ${args.newDelay}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[1],
    },
    {
      address: CROSS_CHAIN_EXECUTOR,
      event: 'event GracePeriodUpdate(uint256 oldGracePeriod, uint256 newGracePeriod)',
      alertId: 'BNB-ADI-GRACE-PERIOD-UPDATED',
      name: '⚠️ BNB a.DI: Cross-chain executor Grace Period updated',
      description: (args: Result) =>
        `Grace Period was updated from ` + `${args.oldGracePeriod} to ${args.newGracePeriod}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[2],
    },
    {
      address: CROSS_CHAIN_EXECUTOR,
      event: 'event MinimumDelayUpdate(uint256 oldMinimumDelay, uint256 newMinimumDelay)',
      alertId: 'BNB-ADI-MIN-DELAY-UPDATED',
      name: '⚠️ BNB a.DI: Cross-chain executor Min Delay updated',
      description: (args: Result) =>
        `Min Delay was updated from ` + `${args.oldMinimumDelay} to ${args.newMinimumDelay}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[3],
    },
    {
      address: CROSS_CHAIN_EXECUTOR,
      event: 'event MaximumDelayUpdate(uint256 oldMaximumDelay, uint256 newMaximumDelay)',
      alertId: 'BNB-ADI-MAX-DELAY-UPDATED',
      name: '⚠️ BNB a.DI: Cross-chain executor Max Delay updated',
      description: (args: Result) =>
        `Max Delay was updated from ` + `${args.oldMaximumDelay} to ${args.newMaximumDelay}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[4],
    },
    {
      address: CROSS_CHAIN_EXECUTOR,
      event:
        'event ActionsSetQueued(uint256 indexed id, address[] targets, uint256[] values, string[] signatures, bytes[] calldatas, bool[] withDelegatecalls, uint256 executionTime)',
      alertId: 'BNB-ADI-ACTION-SET-QUEUED',
      name: 'ℹ️ BNB a.DI: Action set queued',
      description: (args: Result) => `Action set ${args.id} was queued`,
      severity: FindingSeverity.Info,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[5],
    },
    {
      address: CROSS_CHAIN_EXECUTOR,
      event: 'event ActionsSetExecuted(uint256 indexed id, address indexed initiatorExecution, bytes[] returnedData)',
      alertId: 'BNB-ADI-ACTION-SET-EXECUTED',
      name: 'ℹ️ BNB a.DI: Action set executed',
      description: (args: Result) => `Action set ${args.id} was executed`,
      severity: FindingSeverity.Info,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[6],
    },
    {
      address: CROSS_CHAIN_EXECUTOR,
      event: 'event ActionsSetCanceled(uint256 indexed id)',
      alertId: 'BNB-ADI-ACTION-SET-CANCELED',
      name: 'ℹ️ BNB a.DI: Action set canceled',
      description: (args: Result) => `Action set ${args.id} was canceled`,
      severity: FindingSeverity.Info,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[7],
    },
  ]
}
