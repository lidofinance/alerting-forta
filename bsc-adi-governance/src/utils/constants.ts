export const CROSS_CHAIN_EXECUTOR_ADDRESS = '0x8e5175d17f74d1d512de59b2f5d5a5d8177a123d'
export const CROSS_CHAIN_CONTROLLER_ADDRESS = '0x40c4464fca8cacd550c33b39d674fc257966022f'
export const PERIODICAL_BLOCK_INTERVAL = 20 * 5 // 5 min in blocks
export const HOUR_IN_BLOCKS = 20 * 60 // 60 min in blocks

export enum ENVELOPE_STATE {
  NONE,
  CONFIRMED,
  DELIVERED,
}

// ACL
export const NEW_OWNER_IS_CONTRACT_REPORT_INTERVAL = 24 * 60 * 60 // 24h
export const NEW_OWNER_IS_EOA_REPORT_INTERVAL = 60 * 60 // 1h

// Rewards contracts allowed owners
export const WHITELISTED_OWNERS = [CROSS_CHAIN_EXECUTOR_ADDRESS]

export interface IOwnable {
  name: string
  ownershipMethod: string
  ownerAddress?: string
}

// List of contracts to monitor for owner
export const OWNABLE_CONTRACTS = new Map<string, IOwnable>([
  [
    CROSS_CHAIN_CONTROLLER_ADDRESS,
    {
      name: 'BSC ADI upgradeable Proxy Admin',
      ownershipMethod: 'owner',
    },
  ],
])
