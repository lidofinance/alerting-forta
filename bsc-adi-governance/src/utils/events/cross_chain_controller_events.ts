import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { EventOfNotice } from '../types'

export function getCrossChainControllerEvents(crossChainControllerAddress: string): EventOfNotice[] {
  return [
    {
      address: crossChainControllerAddress,
      event: 'event AdminChanged(address previousAdmin, address newAdmin)',
      alertId: 'PROXY-ADMIN-CHANGED',
      name: `ðŸš¨ðŸš¨ðŸš¨ BSC a.DI: Proxy admin changed`,
      description: (args: Result) =>
        `Proxy admin for bnbCrossChainControllerAddress (${crossChainControllerAddress}) ` +
        `was changed from ${args.previousAdmin} to ${args.newAdmin}` +
        `\n(detected by event)`,
      severity: FindingSeverity.Critical,
      type: FindingType.Info,
    },
    {
      address: crossChainControllerAddress,
      event: 'event Upgraded(address indexed implementation)',
      alertId: 'PROXY-UPGRADED',
      name: `ðŸš¨ðŸš¨ðŸš¨ BSC a.DI: Proxy upgraded`,
      description: (args: Result) =>
        `Proxy for bnbCrossChainControllerAddress (${crossChainControllerAddress}) ` +
        `was updated to ${args.implementation}` +
        `\n(detected by event)`,
      severity: FindingSeverity.Critical,
      type: FindingType.Info,
    },
    {
      address: crossChainControllerAddress,
      event: 'event GuardianUpdated(address oldGuardian, address newGuardian)',
      alertId: 'BSC-ADI-GUARDIAN-UPDATED',
      name: 'ðŸš¨ðŸš¨ðŸš¨ BSC a.DI: Guardians updated',
      description: (args: Result) => `Guardian was updated from ` + `${args.oldGuardian} to ${args.newGuardian}`,
      severity: FindingSeverity.Critical,
      type: FindingType.Info,
    },
    {
      address: crossChainControllerAddress,
      event: 'event ConfirmationsUpdated(uint8 newConfirmations, uint256 indexed chainId)',
      alertId: 'BSC-ADI-CONFIRMATIONS-UPDATED',
      name: 'ðŸš¨ BSC a.DI: Allowed Bridges quorum updated',
      description: (args: Result) => `Allowed bridges quorum updated to ${args.newConfirmations}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
    },
    {
      address: crossChainControllerAddress,
      event:
        'event ReceiverBridgeAdaptersUpdated(address indexed bridgeAdapter, bool indexed allowed, uint256 indexed chainId)',
      alertId: 'BSC-ADI-BRIDGE-SET-UPDATED',
      name: 'ðŸš¨ BSC a.DI: Allowed bridges set updated',
      description: (args: Result) =>
        `Allowed bridges set updated - ${args.allowed ? 'added' : 'removed'} ${args.bridgeAdapter}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
    },
  ]
}
