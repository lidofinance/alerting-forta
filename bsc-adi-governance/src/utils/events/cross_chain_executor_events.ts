import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { EventOfNotice } from '../types'

export function getCrossChainExecutorEvents(crossChainExecutorAddress: string): EventOfNotice[] {
  return [
    {
      address: crossChainExecutorAddress,
      event: 'event GuardianUpdate(address oldGuardian, address newGuardian)',
      alertId: 'BSC-ADI-GUARDIAN-UPDATED',
      name: '🚨🚨🚨 BSC a.DI: Guardians updated',
      description: (args: Result) => `Guardian was updated from ` + `${args.oldGuardian} to ${args.newGuardian}`,
      severity: FindingSeverity.Critical,
      type: FindingType.Info,
    },
    {
      address: crossChainExecutorAddress,
      event: 'event DelayUpdate(uint256 oldDelay, uint256 newDelay)',
      alertId: 'BSC-ADI-DELAY-UPDATED',
      name: '⚠️ BSC a.DI: Cross-chain executor delay updated',
      description: (args: Result) => `Delay was updated from ` + `${args.oldDelay} to ${args.newDelay}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
    },
    {
      address: crossChainExecutorAddress,
      event: 'event GracePeriodUpdate(uint256 oldGracePeriod, uint256 newGracePeriod)',
      alertId: 'BSC-ADI-GRACE-PERIOD-UPDATED',
      name: '⚠️ BSC a.DI: Cross-chain executor Grace Period updated',
      description: (args: Result) =>
        `Grace Period was updated from ` + `${args.oldGracePeriod} to ${args.newGracePeriod}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
    },
    {
      address: crossChainExecutorAddress,
      event: 'event MinimumDelayUpdate(uint256 oldMinimumDelay, uint256 newMinimumDelay)',
      alertId: 'BSC-ADI-MIN-DELAY-UPDATED',
      name: '⚠️ BSC a.DI: Cross-chain executor Min Delay updated',
      description: (args: Result) =>
        `Min Delay was updated from ` + `${args.oldMinimumDelay} to ${args.newMinimumDelay}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
    },
    {
      address: crossChainExecutorAddress,
      event: 'event MaximumDelayUpdate(uint256 oldMaximumDelay, uint256 newMaximumDelay)',
      alertId: 'BSC-ADI-MAX-DELAY-UPDATED',
      name: '⚠️ BSC a.DI: Cross-chain executor Max Delay updated',
      description: (args: Result) =>
        `Max Delay was updated from ` + `${args.oldMaximumDelay} to ${args.newMaximumDelay}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
    },
    {
      address: crossChainExecutorAddress,
      event:
        'event ActionsSetQueued(uint256 indexed id, address[] targets, uint256[] values, string[] signatures, bytes[] calldatas, bool[] withDelegatecalls, uint256 executionTime)',
      alertId: 'BSC-ADI-ACTION-SET-QUEUED',
      name: 'ℹ️ BSC a.DI: Action set queued',
      description: (args: Result) => `Action set ${args.id} was queued`,
      severity: FindingSeverity.Info,
      type: FindingType.Info,
    },
    {
      address: crossChainExecutorAddress,
      event: 'event ActionsSetExecuted(uint256 indexed id, address indexed initiatorExecution, bytes[] returnedData)',
      alertId: 'BSC-ADI-ACTION-SET-EXECUTED',
      name: 'ℹ️ BSC a.DI: Action set executed',
      description: (args: Result) => `Action set ${args.id} was executed`,
      severity: FindingSeverity.Info,
      type: FindingType.Info,
    },
    {
      address: crossChainExecutorAddress,
      event: 'event ActionsSetCanceled(uint256 indexed id)',
      alertId: 'BSC-ADI-ACTION-SET-CANCELED',
      name: 'ℹ️ BSC a.DI: Action set canceled',
      description: (args: Result) => `Action set ${args.id} was canceled`,
      severity: FindingSeverity.Info,
      type: FindingType.Info,
    },
  ]
}
