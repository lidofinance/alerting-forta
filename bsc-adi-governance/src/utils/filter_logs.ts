import { ethers } from 'forta-agent'
import { LogDescription, TransactionDto } from './types'

export function filterLogs(txEvent: TransactionDto, eventSignature: string, eventAddress: string): LogDescription[] {
  const iface = new ethers.utils.Interface([eventSignature])
  const filteredLogs: LogDescription[] = []

  for (const log of txEvent.logs) {
    if (log.address.toLowerCase() !== eventAddress.toLowerCase()) {
      continue
    }

    try {
      const logDescription = iface.parseLog(log)
      filteredLogs.push(logDescription)
    } catch (e) {
      // Only one from eventsOfNotice could be correct
      // Others - skipping
    }
  }

  return filteredLogs
}
