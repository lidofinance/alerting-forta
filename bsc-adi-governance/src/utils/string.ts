export function bscscanAddress(address: string): string {
  return `[${address}](https://bscscan.com/address/${address})`
}
