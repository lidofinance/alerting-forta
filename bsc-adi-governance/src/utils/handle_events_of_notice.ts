import { Finding, FindingType } from 'forta-agent'
import { EventOfNotice, TransactionDto } from './types'
import { filterLogs } from './filter_logs'

export function handleEventsOfNotice(txEvent: TransactionDto, eventsOfNotice: EventOfNotice[]) {
  const findings: Finding[] = []

  for (const eventInfo of eventsOfNotice) {
    const filteredLogs = filterLogs(txEvent, eventInfo.event, eventInfo.address)
    for (const filteredLog of filteredLogs) {
      findings.push(
        Finding.fromObject({
          name: eventInfo.name,
          description: eventInfo.description(filteredLog.args),
          alertId: eventInfo.alertId,
          severity: eventInfo.severity,
          type: FindingType.Info,
          metadata: { args: String(filteredLog.args) },
        }),
      )
    }
  }

  return findings
}
