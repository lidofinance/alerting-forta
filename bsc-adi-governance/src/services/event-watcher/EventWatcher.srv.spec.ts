import * as Winston from 'winston'
import { ethers, Finding } from 'forta-agent'

import { EventWatcherSrv } from './EventWatcher.srv'
import { getFortaConfig } from 'forta-agent/dist/sdk/utils'
import { getCrossChainExecutorEvents } from '../../utils/events/cross_chain_executor_events'
import { CROSS_CHAIN_EXECUTOR_ADDRESS } from '../../utils/constants'
import { TransactionDto } from '../../utils/types'
import { TransactionEvent } from 'forta-agent/dist/sdk/transaction.event'

const TEST_TIMEOUT = 120_000 // ms

describe('EventWatcher srv functional tests', () => {
  const chainId = 56
  const ethProvider = new ethers.providers.JsonRpcProvider(getFortaConfig().jsonRpcUrl, chainId)

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const service = new EventWatcherSrv(
    'CrossChainExecutorWatcher',
    getCrossChainExecutorEvents(CROSS_CHAIN_EXECUTOR_ADDRESS),
    logger,
  )

  let runTransaction: (txHash: string, initBlock?: number) => Promise<Finding[]>

  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation(() => {})
    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2024-01-01').getTime())

    runTransaction = async (txHash: string) => {
      const trx = await ethProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
      }

      return service.handleTransaction(transactionDto as TransactionEvent)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should find nothing',
    async () => {
      const findings = await runTransaction(
        '0xda1572eb6360116ea0fbd9f898b1413823ba2ab4bb1cbb5e150cce42c68a687c',
        39775722,
      )
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
