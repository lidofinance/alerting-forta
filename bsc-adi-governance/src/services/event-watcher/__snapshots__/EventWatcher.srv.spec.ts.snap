// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventWatcher srv functional tests should find nothing 1`] = `[]`;
